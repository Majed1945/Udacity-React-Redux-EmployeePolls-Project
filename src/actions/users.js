export const RECEIVE_USERS = "RECEIVE_USERS";
export const SET_USER_ANSWER = "SET_USER_ANSWER";
export const ASSIGN_QUESTION = "ASSIGN_QUESTION";
export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}
export function setUserAnswer(qid, authedUser, answer) {
  return {
    type: SET_USER_ANSWER,
    qid,
    authedUser,
    answer,
  };
}
export function assignQuestion(author, questionId) {
  return {
    type: ASSIGN_QUESTION,
    author,
    questionId,
  };
}
