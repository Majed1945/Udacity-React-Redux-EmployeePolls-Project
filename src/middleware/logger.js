const logger = (store) => (next) => (action) => {
    console.group(action.type);
    console.log("The action is ", action);
    //here we dispatch the action
    const returnValue = next(action);
    //after, the state will be updated
    console.log("The new state is ", store.getState());
    console.groupEnd();
    return returnValue;
  };
  export default logger;
  