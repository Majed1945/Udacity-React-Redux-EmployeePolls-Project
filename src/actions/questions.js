import { saveQuestionAnswer, saveQuestion } from "../utils/api";
import { setUserAnswer,assignQuestion } from "./users";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ANSWER_QUESTION = "ANSWER_QUESTION";
export const ADD_QUESTION = "ADD_QUESTION";
export const CREATE_QUESTION = "CREATE_QUESTION";
export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

function answerQuestion(qid, authedUser, answer) {
  return {
    type: ANSWER_QUESTION,
    qid,
    authedUser,
    answer,
  };
}

export function handleAnswerQuestion(qid, authedUser, answer) {
  return (dispatch) => {
    return saveQuestionAnswer(qid, authedUser, answer)
      .catch((e) => {
        console.warn("Error in handleAnswerQuestion ", e);
        alert("The was an error saving the question answer. Try again.");
      })
      .then(() => {
        dispatch(answerQuestion(qid, authedUser, answer));
        dispatch(setUserAnswer(qid, authedUser, answer));
      });
  };
}
function createQuestion(question) {
  return {
    type: CREATE_QUESTION,
    question,
  };
}

export function handleCreateQuestion(question) {
  return (dispatch) => {
    return saveQuestion(question)
      .catch((e) => {
        console.warn("Error in saving new question ", e);
        alert("The was an error saving the question . Try again.");
      })
      .then((question) => {
        dispatch(createQuestion(question));
        dispatch(assignQuestion(question.author,question.id));
      });
  };
}
