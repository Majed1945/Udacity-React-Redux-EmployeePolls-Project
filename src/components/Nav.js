import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import profile from "../images/profile.png";
import { useNavigate } from "react-router-dom";
const Nav = (props) => {
  const navigate = useNavigate();
  const currentUser = props.users.filter((user) => {
    return user.id === props.authedUser;
  });

  function logOut() {
    props.dispatch(setAuthedUser(null));
    navigate("/");
  }
  return (
    <nav className="p-1 sticky top-0 z-10 backdrop-filter backdrop-blur-lg bg-opacity-30 border-b border-gray-200 bg-white dark:bg-gray-800 dark:border-gray-700">
      <div className="bg-transparent container flex flex-wrap items-center justify-between ml-2">
        <button
          data-collapse-toggle="navbar-default"
          type="button"
          className="inline-flex items-center p-2 ml-3 text-sm text-gray-500 rounded-lg md:hidden hover:bg-gray-100 focus:outline-none focus:ring-2 focus:ring-gray-200 dark:text-gray-400 dark:hover:bg-gray-700 dark:focus:ring-gray-600"
          aria-controls="navbar-default"
          aria-expanded="false"
        >
          <span className="sr-only">Open main menu</span>
          <svg
            className="w-6 h-6"
            aria-hidden="true"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              fillRule="evenodd"
              d="M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 15a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
              clipRule="evenodd"
            ></path>
          </svg>
        </button>
        <div className="hidden w-full md:block md:w-auto" id="navbar-default">
          <ul className="flex flex-col items-center p-3 mt-4 border rounded-lg bg-transparent md:flex-row md:space-x-8 md:mt-0 md:text-sm md:font-medium md:border-0  dark:bg-gray-800 md:dark:bg-gray-900 dark:border-gray-700">
            <li className="block">
              {props.authedUser !== null ? (
                <div className="flex flex-wrap items-center">
                  <p className="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent">
                    {currentUser[0].name}
                  </p>
                  <img className="w-12" src={currentUser[0].avatarURL} />
                </div>
              ) : (
                <>
                  <img className="w-10" src={profile} />
                </>
              )}
            </li>

            <li>
              <Link
                to="/"
                className="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
                aria-current="page"
              >
                Home
              </Link>
            </li>
            <li>
              <Link
                to="/add"
                className="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
              >
                New Question
              </Link>
            </li>
            <li>
              <Link
                to="/leaderboard"
                className="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
              >
                Leaderboard
              </Link>
            </li>

            <li className=" position: absolute right-16">
              {props.authedUser !== null ? (
                <button
                  data-testid="nav-sign-out"
                  onClick={logOut}
                  className="bg-black text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
                  type="button"
                >
                  Log Out
                </button>
              ) : (
                <></>
              )}
            </li>
          </ul>
        </div>
      </div>
    </nav>
  );
};
const mapStateToProps = ({ authedUser, users }) => ({
  authedUser,
  users: Object.values(users),
});
export default connect(mapStateToProps)(Nav);
