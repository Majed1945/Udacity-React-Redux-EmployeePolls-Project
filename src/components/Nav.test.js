import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";
import React from "react";
import { BrowserRouter } from "react-router-dom";
import Nav from "./Nav";
import { createStore } from "redux";
import reducer from "../reducers";
import middleware from "../middleware";
import { setAuthedUser } from "../actions/authedUser";
import { handleInitialData } from "../actions/shared";

const store = createStore(reducer, middleware);

describe("Nav", () => {
  it("should render the nav bar correctly", () => {
    const renderedNav = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );
    expect(renderedNav).toMatchSnapshot();
  });
  it("should not display the Log Out button when the user is logged out", async () => {
    const nav = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );
    expect(nav.queryByTestId("nav-sign-out")).not.toBeInTheDocument();
  });
  it("should display the Log Out button when the user is logged in", async () => {
    await store.dispatch(handleInitialData());
    store.dispatch(setAuthedUser("sarahedo"));
    const nav = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );
    expect(nav.getByTestId("nav-sign-out")).toBeInTheDocument();
  });
  it("when the user is logged in and presses the log out botton, the log out button should disappear.", async () => {
    await store.dispatch(handleInitialData());
    store.dispatch(setAuthedUser("sarahedo"));
    const nav = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );
    const signOutButton = nav.getByTestId("nav-sign-out");
    fireEvent.click(signOutButton);
    expect(nav.queryByTestId("nav-sign-out")).not.toBeInTheDocument();
  });
});
