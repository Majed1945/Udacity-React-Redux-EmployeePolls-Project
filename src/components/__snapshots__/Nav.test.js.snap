// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should render the nav bar correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="p-1 sticky top-0 z-10 backdrop-filter backdrop-blur-lg bg-opacity-30 border-b border-gray-200 bg-white dark:bg-gray-800 dark:border-gray-700"
      >
        <div
          class="bg-transparent container flex flex-wrap items-center justify-between ml-2"
        >
          <button
            aria-controls="navbar-default"
            aria-expanded="false"
            class="inline-flex items-center p-2 ml-3 text-sm text-gray-500 rounded-lg md:hidden hover:bg-gray-100 focus:outline-none focus:ring-2 focus:ring-gray-200 dark:text-gray-400 dark:hover:bg-gray-700 dark:focus:ring-gray-600"
            data-collapse-toggle="navbar-default"
            type="button"
          >
            <span
              class="sr-only"
            >
              Open main menu
            </span>
            <svg
              aria-hidden="true"
              class="w-6 h-6"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 15a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <div
            class="hidden w-full md:block md:w-auto"
            id="navbar-default"
          >
            <ul
              class="flex flex-col items-center p-3 mt-4 border rounded-lg bg-transparent md:flex-row md:space-x-8 md:mt-0 md:text-sm md:font-medium md:border-0  dark:bg-gray-800 md:dark:bg-gray-900 dark:border-gray-700"
            >
              <li
                class="block"
              >
                <img
                  class="w-10"
                  src="profile.png"
                />
              </li>
              <li>
                <a
                  aria-current="page"
                  class="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
                  href="/"
                >
                  Home
                </a>
              </li>
              <li>
                <a
                  class="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
                  href="/add"
                >
                  New Question
                </a>
              </li>
              <li>
                <a
                  class="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </li>
              <li
                class=" position: absolute right-16"
              />
            </ul>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="p-1 sticky top-0 z-10 backdrop-filter backdrop-blur-lg bg-opacity-30 border-b border-gray-200 bg-white dark:bg-gray-800 dark:border-gray-700"
    >
      <div
        class="bg-transparent container flex flex-wrap items-center justify-between ml-2"
      >
        <button
          aria-controls="navbar-default"
          aria-expanded="false"
          class="inline-flex items-center p-2 ml-3 text-sm text-gray-500 rounded-lg md:hidden hover:bg-gray-100 focus:outline-none focus:ring-2 focus:ring-gray-200 dark:text-gray-400 dark:hover:bg-gray-700 dark:focus:ring-gray-600"
          data-collapse-toggle="navbar-default"
          type="button"
        >
          <span
            class="sr-only"
          >
            Open main menu
          </span>
          <svg
            aria-hidden="true"
            class="w-6 h-6"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 15a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <div
          class="hidden w-full md:block md:w-auto"
          id="navbar-default"
        >
          <ul
            class="flex flex-col items-center p-3 mt-4 border rounded-lg bg-transparent md:flex-row md:space-x-8 md:mt-0 md:text-sm md:font-medium md:border-0  dark:bg-gray-800 md:dark:bg-gray-900 dark:border-gray-700"
          >
            <li
              class="block"
            >
              <img
                class="w-10"
                src="profile.png"
              />
            </li>
            <li>
              <a
                aria-current="page"
                class="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <a
                class="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
                href="/add"
              >
                New Question
              </a>
            </li>
            <li>
              <a
                class="block py-2 pl-3 pr-4 text-gray-700 rounded hover:bg-gray-100 md:hover:bg-transparent md:border-0 md:hover:text-blue-700 md:p-0 dark:text-gray-400 md:dark:hover:text-white dark:hover:bg-gray-700 dark:hover:text-white md:dark:hover:bg-transparent"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li
              class=" position: absolute right-16"
            />
          </ul>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
